File: sdks/php/src/InvokeilPay.php
<?php
/**
 * Invokeil Pay — official PHP SDK (single file, no dependencies, curl only).
 *
 *   $invokeil = new InvokeilPay('https://pay.example.com', 'sk_live_xxx');
 *   $checkout = $invokeil->createCheckout([...]);
 *
 * Compatible with PHP 7.2+.
 */

if (!class_exists('InvokeilPay')) {

class InvokeilPay
{
    /** @var string */
    private $baseUrl;
    /** @var string */
    private $apiKey;
    /** @var int */
    private $timeout;

    /**
     * @param string $baseUrl e.g. https://pay.example.com
     * @param string $apiKey  store API key (sk_live_… / sk_test_…)
     * @param int    $timeout request timeout in seconds
     */
    public function __construct($baseUrl, $apiKey, $timeout = 30)
    {
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->apiKey = $apiKey;
        $this->timeout = $timeout;
    }

    /**
     * Create a hosted checkout (payment).
     *
     * @param array $payment { amount, customer_name?, customer_mobile?, customer_email?,
     *                        redirect_url?, cancel_url?, webhook_url?, metadata? }
     * @return array decoded API response
     * @throws RuntimeException on HTTP or API error
     */
    public function createCheckout(array $payment)
    {
        return $this->request('POST', '/api/v1/checkout', $payment);
    }

    /** PipraPay-compatible alias of createCheckout(). */
    public function createPayment(array $payment)
    {
        return $this->createCheckout($payment);
    }

    /**
     * Verify a payment by checkout id (PipraPay-style pp_id).
     *
     * @param string $ppId
     * @return array decoded API response
     */
    public function verifyPayment($ppId)
    {
        return $this->request('POST', '/api/v1/verify-payment', ['pp_id' => (string) $ppId]);
    }

    /**
     * Fetch a checkout by token.
     *
     * @param string $token
     * @return array decoded API response
     */
    public function getCheckout($token)
    {
        return $this->request('GET', '/api/v1/checkout/' . rawurlencode($token));
    }

    /**
     * Verify a webhook signature produced by Invokeil Pay.
     * Scheme: header "t=<ms-epoch>,v1=hex(hmac_sha256(secret, '<t>.<rawBody>'))"
     * with a 5-minute replay window.
     *
     * @param string $rawBody exact request body (file_get_contents('php://input'))
     * @param string $header  the X-Invokeil-Signature header value
     * @param string $secret  the store's webhook signing secret (whsec_…)
     * @param int    $toleranceMs replay window in milliseconds (default 300000)
     * @return bool
     */
    public function verifyWebhookSignature($rawBody, $header, $secret, $toleranceMs = 300000)
    {
        if (!$rawBody || !$header || !$secret) {
            return false;
        }
        $parts = [];
        foreach (explode(',', (string) $header) as $kv) {
            $i = strpos($kv, '=');
            if ($i > 0) {
                $parts[trim(substr($kv, 0, $i))] = trim(substr($kv, $i + 1));
            }
        }
        if (empty($parts['t']) || empty($parts['v1'])) {
            return false;
        }
        $t = $parts['t'];
        if (!ctype_digit($t) || abs((time() * 1000) - (int) $t) > $toleranceMs) {
            return false;
        }
        $expected = hash_hmac('sha256', $t . '.' . $rawBody, $secret);
        return hash_equals($expected, $parts['v1']);
    }

    /**
     * Internal HTTP helper via curl.
     *
     * @return array decoded JSON
     * @throws RuntimeException
     */
    private function request($method, $path, array $body = null)
    {
        if (!function_exists('curl_init')) {
            throw new RuntimeException('curl extension is required');
        }
        $ch = curl_init($this->baseUrl . $path);
        $headers = [
            'Authorization: Bearer ' . $this->apiKey,
            'Accept: application/json',
        ];
        if ($method === 'POST') {
            curl_setopt($ch, CURLOPT_POST, true);
            curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($body ?? new stdClass()));
            $headers[] = 'Content-Type: application/json';
        }
        curl_setopt_array($ch, [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER     => $headers,
            CURLOPT_TIMEOUT        => $this->timeout,
            CURLOPT_CONNECTTIMEOUT => 10,
        ]);
        $res = curl_exec($ch);
        if ($res === false) {
            $err = curl_error($ch);
            curl_close($ch);
            throw new RuntimeException('HTTP request failed: ' . $err);
        }
        $status = (int) curl_getinfo($ch, CURLINFO_RESPONSE_CODE);
        curl_close($ch);

        $data = json_decode((string) $res, true);
        if (!is_array($data)) {
            throw new RuntimeException('Invalid JSON response (HTTP ' . $status . ')');
        }
        if ($status >= 400) {
            $msg = isset($data['error']) ? $data['error'] : ('HTTP ' . $status);
            throw new RuntimeException('Invokeil Pay API error: ' . $msg, $status);
        }
        return $data;
    }
}

}

File: sdks/dotnet/InvokeilPay.cs
// Invokeil Pay — official .NET / C# SDK (single file, HttpClient only).
//
// Usage:
//   var invokeil = new InvokeilPay("https://pay.example.com", "sk_live_xxx");
//   var checkout = await invokeil.CreateCheckoutAsync(new { amount = 500, customer_mobile = "01712345678" });
//
// Targets .NET Standard 2.0+ / .NET 6+ (System.Text.Json).
#nullable enable
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace Invokeil.Pay
{
    public class InvokeilPayException : Exception
    {
        public int Status { get; }
        public string? Body { get; }

        public InvokeilPayException(string message, int status = 0, string? body = null)
            : base(message)
        {
            Status = status;
            Body = body;
        }
    }

    public class InvokeilPay
    {
        private readonly HttpClient _http;
        private static readonly JsonSerializerOptions JsonOpts = new()
        {
            PropertyNameCaseInsensitive = true,
        };

        /// <param name="baseUrl">e.g. https://pay.example.com</param>
        /// <param name="apiKey">store API key (sk_live_… / sk_test_…)</param>
        /// <param name="httpClient">optional HttpClient (a default one is created otherwise)</param>
        public InvokeilPay(string baseUrl, string apiKey, HttpClient? httpClient = null)
        {
            if (string.IsNullOrWhiteSpace(baseUrl)) throw new ArgumentException("baseUrl is required");
            if (string.IsNullOrWhiteSpace(apiKey)) throw new ArgumentException("apiKey is required");
            BaseUrl = baseUrl.TrimEnd('/');
            _http = httpClient ?? new HttpClient { Timeout = TimeSpan.FromSeconds(30) };
            _http.DefaultRequestHeaders.Remove("Authorization");
            _http.DefaultRequestHeaders.TryAddWithoutValidation("Authorization", $"Bearer {apiKey}");
            _http.DefaultRequestHeaders.TryAddWithoutValidation("Accept", "application/json");
        }

        public string BaseUrl { get; }

        /// <summary>Create a hosted checkout. POST /api/v1/checkout</summary>
        public Task<JsonElement> CreateCheckoutAsync(object payment) =>
            SendAsync(HttpMethod.Post, "/api/v1/checkout", payment);

        /// <summary>PipraPay-compatible alias of CreateCheckoutAsync.</summary>
        public Task<JsonElement> CreatePaymentAsync(object payment) =>
            CreateCheckoutAsync(payment);

        /// <summary>Verify a payment by checkout id (PipraPay-style pp_id).</summary>
        public Task<JsonElement> VerifyPaymentAsync(string ppId) =>
            SendAsync(HttpMethod.Post, "/api/v1/verify-payment", new { pp_id = ppId });

        /// <summary>Fetch a checkout by token. GET /api/v1/checkout/{token}</summary>
        public Task<JsonElement> GetCheckoutAsync(string token) =>
            SendAsync(HttpMethod.Get, $"/api/v1/checkout/{Uri.EscapeDataString(token)}", null);

        /// <summary>
        /// Verify an X-Invokeil-Signature header.
        /// Scheme: t=&lt;ms-epoch&gt;,v1=hex(hmac_sha256(secret, "&lt;t&gt;.&lt;rawBody&gt;")), 5-minute replay window.
        /// Pass the RAW body exactly as received — never a re-serialized object.
        /// </summary>
        public static bool VerifyWebhookSignature(string rawBody, string header, string secret, int toleranceMs = 300_000)
        {
            if (string.IsNullOrEmpty(rawBody) || string.IsNullOrEmpty(header) || string.IsNullOrEmpty(secret))
                return false;

            var parts = new Dictionary<string, string>();
            foreach (var kv in header.Split(','))
            {
                int i = kv.IndexOf('=');
                if (i > 0)
                    parts[kv.Substring(0, i).Trim()] = kv.Substring(i + 1).Trim();
            }
            if (!parts.TryGetValue("t", out var t) || !parts.TryGetValue("v1", out var v1))
                return false;
            if (!long.TryParse(t, NumberStyles.Integer, CultureInfo.InvariantCulture, out var ts))
                return false;
            if (Math.Abs(DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - ts) > toleranceMs)
                return false;

            using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
            byte[] mac = hmac.ComputeHash(Encoding.UTF8.GetBytes($"{t}.{rawBody}"));
            var expected = Convert.ToHexString(mac).ToLowerInvariant();
            return FixedTimeEquals(expected, v1.ToLowerInvariant());
        }

        private static bool FixedTimeEquals(string a, string b)
        {
            if (a.Length != b.Length) return false;
            int diff = 0;
            for (int i = 0; i < a.Length; i++) diff |= a[i] ^ b[i];
            return diff == 0;
        }

        private async Task<JsonElement> SendAsync(HttpMethod method, string path, object? body)
        {
            using var req = new HttpRequestMessage(method, BaseUrl + path);
            if (body is not null)
            {
                req.Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");
            }
            using var res = await _http.SendAsync(req);
            string text = await res.Content.ReadAsStringAsync();
            JsonElement data = JsonDocument.Parse(string.IsNullOrWhiteSpace(text) ? "{}" : text).RootElement.Clone();
            if (!res.IsSuccessStatusCode)
            {
                string msg = data.ValueKind == JsonValueKind.Object && data.TryGetProperty("error", out var e)
                    ? e.GetString() ?? $"HTTP {(int)res.StatusCode}"
                    : $"HTTP {(int)res.StatusCode}";
                throw new InvokeilPayException(msg, (int)res.StatusCode, text);
            }
            return data;
        }
    }
}
